File: python/audio/g711.py
from __future__ import annotations

import numpy as np

ULAW_BIAS: int = 0x84
ULAW_CLIP: int = 32_635

def _leading_seg_8bit(v: int) -> int:
    v &= 0xFF
    if v & 0x80:
        return 7
    if v & 0x40:
        return 6
    if v & 0x20:
        return 5
    if v & 0x10:
        return 4
    if v & 0x08:
        return 3
    if v & 0x04:
        return 2
    if v & 0x02:
        return 1
    return 0

def _to_i16(v: int) -> int:
    v &= 0xFFFF
    if v >= 0x8000:
        v -= 0x10000
    return v

def ulaw_decode_byte(u: int) -> int:
    u = (~u) & 0xFF
    sign = u & 0x80
    exponent = (u >> 4) & 0x07
    mantissa = u & 0x0F
    t = (mantissa << 3) + ULAW_BIAS
    t <<= exponent
    s = (ULAW_BIAS - t) if sign != 0 else (t - ULAW_BIAS)
    return _to_i16(s)

def ulaw_encode_sample(s: int) -> int:
    sample = int(s)
    if sample < 0:
        sample = -sample
        sign = 0x80
    else:
        sign = 0
    if sample > ULAW_CLIP:
        sample = ULAW_CLIP
    sample += ULAW_BIAS
    exponent = _leading_seg_8bit((sample >> 7) & 0xFF)
    mantissa = (sample >> (exponent + 3)) & 0x0F
    return (~(sign | (exponent << 4) | mantissa)) & 0xFF

def alaw_decode_byte(a: int) -> int:
    a = (a ^ 0x55) & 0xFF
    mantissa = a & 0x0F
    seg = (a & 0x70) >> 4
    t = mantissa << 4
    if seg == 0:
        t += 8
    elif seg == 1:
        t += 0x108
    else:
        t += 0x108
        t <<= seg - 1
    val = t if (a & 0x80) != 0 else -t
    if val > 32_767:
        val = 32_767
    elif val < -32_768:
        val = -32_768
    return val

_ALAW_SEG_END = (0xFF, 0x1FF, 0x3FF, 0x7FF, 0xFFF, 0x1FFF, 0x3FFF, 0x7FFF)

def alaw_encode_sample(s: int) -> int:
    pcm = int(s)
    if pcm >= 0:
        mask = 0xD5
    else:
        pcm = -pcm - 8
        mask = 0x55
    seg = 8
    for i, end in enumerate(_ALAW_SEG_END):
        if pcm <= end:
            seg = i
            break
    if seg >= 8:
        return (0x7F ^ mask) & 0xFF
    aval = (seg & 0xFF) << 4
    if seg < 2:
        mantissa = (pcm >> 4) & 0x0F
    else:
        mantissa = (pcm >> (seg + 3)) & 0x0F
    return ((aval | mantissa) ^ mask) & 0xFF

_ULAW_DECODE_TABLE = np.array(
    [ulaw_decode_byte(i) for i in range(256)], dtype=np.int16
)
_ALAW_DECODE_TABLE = np.array(
    [alaw_decode_byte(i) for i in range(256)], dtype=np.int16
)

_ENCODE_LUT_INDICES = np.arange(65536, dtype=np.uint16).view(np.int16)
_ULAW_ENCODE_LUT = np.array(
    [ulaw_encode_sample(int(x)) for x in _ENCODE_LUT_INDICES], dtype=np.uint8
)
_ALAW_ENCODE_LUT = np.array(
    [alaw_encode_sample(int(x)) for x in _ENCODE_LUT_INDICES], dtype=np.uint8
)

def ulaw_bytes_to_f32(b: bytes | bytearray | memoryview | np.ndarray) -> np.ndarray:
    arr = np.frombuffer(bytes(b), dtype=np.uint8)
    return (_ULAW_DECODE_TABLE[arr].astype(np.float32)) / np.float32(32768.0)

def alaw_bytes_to_f32(b: bytes | bytearray | memoryview | np.ndarray) -> np.ndarray:
    arr = np.frombuffer(bytes(b), dtype=np.uint8)
    return (_ALAW_DECODE_TABLE[arr].astype(np.float32)) / np.float32(32768.0)

def f32_to_ulaw_bytes(samples: np.ndarray) -> bytes:
    s = np.asarray(samples, dtype=np.float32)
    s = np.clip(s, -1.0, 1.0)
    v = np.rint(s * 32767.0).astype(np.int32)
    v = np.clip(v, -32768, 32767).astype(np.int16)
    return _ULAW_ENCODE_LUT[v.view(np.uint16)].tobytes()

def f32_to_alaw_bytes(samples: np.ndarray) -> bytes:
    s = np.asarray(samples, dtype=np.float32)
    s = np.clip(s, -1.0, 1.0)
    v = np.rint(s * 32767.0).astype(np.int32)
    v = np.clip(v, -32768, 32767).astype(np.int16)
    return _ALAW_ENCODE_LUT[v.view(np.uint16)].tobytes()

__all__ = [
    "ULAW_BIAS",
    "ULAW_CLIP",
    "ulaw_decode_byte",
    "ulaw_encode_sample",
    "alaw_decode_byte",
    "alaw_encode_sample",
    "ulaw_bytes_to_f32",
    "alaw_bytes_to_f32",
    "f32_to_ulaw_bytes",
    "f32_to_alaw_bytes",
]

File: rust/src/audio/g711.rs
const ULAW_BIAS: i32 = 0x84;
const ULAW_CLIP: i32 = 32_635;

pub fn ulaw_decode_byte(u: u8) -> i16 {
    let u = !u;
    let sign = u & 0x80;
    let exponent = (u >> 4) & 0x07;
    let mantissa = u & 0x0F;
    let mut t = ((mantissa as i32) << 3) + ULAW_BIAS;
    t <<= exponent;
    let s = if sign != 0 {
        ULAW_BIAS - t
    } else {
        t - ULAW_BIAS
    };
    s as i16
}

pub fn ulaw_encode_sample(s: i16) -> u8 {
    let mut sample = s as i32;
    let sign = if sample < 0 {
        sample = -sample;
        0x80u8
    } else {
        0
    };
    if sample > ULAW_CLIP {
        sample = ULAW_CLIP;
    }
    sample += ULAW_BIAS;
    let exponent = leading_seg_8bit((sample >> 7) as u32 & 0xFF);
    let mantissa = ((sample >> (exponent + 3)) & 0x0F) as u8;
    !(sign | (exponent << 4) | mantissa)
}

fn leading_seg_8bit(v: u32) -> u8 {
    let v = v & 0xFF;
    if v & 0x80 != 0 {
        7
    } else if v & 0x40 != 0 {
        6
    } else if v & 0x20 != 0 {
        5
    } else if v & 0x10 != 0 {
        4
    } else if v & 0x08 != 0 {
        3
    } else if v & 0x04 != 0 {
        2
    } else if v & 0x02 != 0 {
        1
    } else {
        0
    }
}

pub fn alaw_decode_byte(a: u8) -> i16 {
    let a = a ^ 0x55;
    let mantissa = (a & 0x0F) as i32;
    let seg = ((a & 0x70) >> 4) as i32;
    let mut t = mantissa << 4;
    if seg == 0 {
        t += 8;
    } else if seg == 1 {
        t += 0x108;
    } else {
        t += 0x108;
        t <<= seg - 1;
    }
    let val = if a & 0x80 != 0 { t } else { -t };
    val.clamp(i16::MIN as i32, i16::MAX as i32) as i16
}

pub fn alaw_encode_sample(s: i16) -> u8 {
    let mut pcm = s as i32;
    let mask: u8 = if pcm >= 0 {
        0xD5
    } else {
        pcm = -pcm - 8;
        0x55
    };
    static SEG_END: [i32; 8] = [0xFF, 0x1FF, 0x3FF, 0x7FF, 0xFFF, 0x1FFF, 0x3FFF, 0x7FFF];
    let mut seg = 8usize;
    for (i, &end) in SEG_END.iter().enumerate() {
        if pcm <= end {
            seg = i;
            break;
        }
    }
    if seg >= 8 {
        return 0x7F ^ mask;
    }
    let aval = (seg as u8) << 4;
    let mantissa = if seg < 2 {
        ((pcm >> 4) & 0x0F) as u8
    } else {
        ((pcm >> (seg + 3)) & 0x0F) as u8
    };
    (aval | mantissa) ^ mask
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ulaw_decode_is_total_over_byte_space() {
        for b in 0u16..=0xFF {
            let v = ulaw_decode_byte(b as u8) as i32;
            assert!(v.abs() <= 32_124, "ulaw_decode({b:02x}) = {v} out of range");
        }
    }

    #[test]
    fn alaw_decode_is_total_over_byte_space() {
        for b in 0u16..=0xFF {
            let v = alaw_decode_byte(b as u8) as i32;
            assert!(v.abs() <= 32_256, "alaw_decode({b:02x}) = {v} out of range");
        }
    }

    #[test]
    fn ulaw_encode_is_total_over_i16_space() {
        let mut max_rel_err = 0.0f32;
        for s in i16::MIN..=i16::MAX {
            if s == i16::MIN {
                continue;
            }
            let encoded = ulaw_encode_sample(s);
            let decoded = ulaw_decode_byte(encoded);
            if (s.abs() as i32) > 256 {
                if decoded != 0 {
                    assert_eq!(decoded.signum(), s.signum(), "sign for {s}");
                }
                let err = (decoded as f32 - s as f32).abs() / s.abs() as f32;
                if err > max_rel_err {
                    max_rel_err = err;
                }
            }
        }
        assert!(
            max_rel_err < 0.13,
            "max relative error {max_rel_err} exceeds μ-law spec"
        );
    }

    #[test]
    fn alaw_encode_is_total_over_i16_space() {
        let mut max_rel_err = 0.0f32;
        for s in i16::MIN..=i16::MAX {
            let encoded = alaw_encode_sample(s);
            let decoded = alaw_decode_byte(encoded);

            if s == i16::MIN {
                continue;
            }
            if (s.abs() as i32) > 256 {
                if decoded != 0 {
                    assert_eq!(decoded.signum(), s.signum(), "sign for {s}");
                }
                let err = (decoded as f32 - s as f32).abs() / s.abs() as f32;
                if err > max_rel_err {
                    max_rel_err = err;
                }
            }
        }
        assert!(max_rel_err < 0.13, "max relative error {max_rel_err}");
    }

    #[test]
    fn ulaw_round_trip_preserves_sign_and_magnitude() {
        for s in [-32_000i16, -8_000, -1_000, 0, 1_000, 8_000, 32_000] {
            let encoded = ulaw_encode_sample(s);
            let decoded = ulaw_decode_byte(encoded);
            if s == 0 {
                assert!(decoded.abs() <= ULAW_BIAS as i16 / 2);
            } else {
                assert_eq!(decoded.signum(), s.signum(), "sign for {s}");
                let err = (decoded as f32 - s as f32).abs() / s.abs() as f32;
                assert!(err < 0.13, "relative error {err} for {s}");
            }
        }
    }

    #[test]
    fn alaw_round_trip_preserves_sign_and_magnitude() {
        for s in [-32_000i16, -8_000, -1_000, 0, 1_000, 8_000, 32_000] {
            let encoded = alaw_encode_sample(s);
            let decoded = alaw_decode_byte(encoded);
            if s == 0 {
                assert!(decoded.abs() <= 16);
            } else {
                assert_eq!(decoded.signum(), s.signum(), "sign for {s}");
                let err = (decoded as f32 - s as f32).abs() / s.abs() as f32;
                assert!(err < 0.13, "relative error {err} for {s}");
            }
        }
    }

    #[test]
    fn ulaw_silence_decodes_to_zero() {
        assert!(ulaw_decode_byte(0xFF).abs() < 16);
    }

    #[test]
    fn alaw_silence_decodes_to_zero() {
        assert!(alaw_decode_byte(0xD5).abs() < 16);
    }

    #[test]
    fn alaw_extremes_clip_in_max_segment() {
        let pos = alaw_encode_sample(32_767);
        let neg = alaw_encode_sample(-32_768);

        let dp = alaw_decode_byte(pos) as i32;
        let dn = alaw_decode_byte(neg) as i32;
        assert!(dp > 24_000, "positive extreme decodes to {dp}");
        assert!(dn < -24_000, "negative extreme decodes to {dn}");
    }
}
